Introduce GOSSIP log level to PeerHandler

tnull opened the pull request:
As requested in #1140, this PR introduces a new `GOSSIP` log level for gossip messages. Thereby, it should make logs more readable at the `TRACE` level. 

Note that I left the logging events concerning connection management at the `TRACE` level, because I figured they may still be of interest (and are likely not cluttering the logs too much?).

TheBlueMatt commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -707,7 +707,7 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 		let encoded_message = buffer.0;
 
-		log_trace!(self.logger, "Enqueueing message {:?} to {}", message, log_pubkey!(peer.their_node_id.unwrap()));
+		log_gossip!(self.logger, "Enqueueing message {:?} to {}", message, log_pubkey!(peer.their_node_id.unwrap()));
```
Right, so GOSSIP-vs-TRACE is mostly a "messages that we send a *ton*" vs "messages that wont fill your disk up during normal/non-adversarial operation", in my mind.

TheBlueMatt replied:
So basically the only messages that are GOSSIP are things about the channels in the graph that we process via the `RouteHandler` and not messages that get passed to the `channel_handler` later down in `handle_message.

tnull replied:
Okay! This should be addressed by https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/fbbaeea3afd1f02da29c4b5e7981807b806973ff.

TheBlueMatt commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -1391,7 +1391,7 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 							},
 							msgs::ErrorAction::SendErrorMessage { ref msg } => {
-								log_trace!(self.logger, "Handling SendErrorMessage HandleError event in peer_handler for node {} with message {}",
+								log_gossip!(self.logger, "Handling SendErrorMessage HandleError event in peer_handler for node {} with message {}",
```
I think we only want to convert the error handling to things that come from normal gossip operation, which is basically just "we don't support zlib" or "we have already received that message", not things where we send an "error" message, which is usually "something is very bad, and we're closing a channel with you"

tnull replied:
Let me know if I'm mistaken, but as far as I see right now there is no good way to discern the different `ErrorMessage` types? 
I therefore reverted this back to `log_trace` for the time being with https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/fbbaeea3afd1f02da29c4b5e7981807b806973ff.

TheBlueMatt replied:
Generally, indeed, long-term we want to convert the `IgnoreError` types to `IgnoreAndLog` (mostly), which lets the originator of the error give more instruction here.

TheBlueMatt reviewed (commented):
So Sorry for the delay here, been an insane week trying to get 0.0.103 out the door. Things should be back to normal mostly now :)

tnull commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -901,7 +901,7 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 		message: wire::Message<<<CMH as core::ops::Deref>::Target as wire::CustomMessageReader>::CustomMessage>
 	) -> Result<Option<wire::Message<<<CMH as core::ops::Deref>::Target as wire::CustomMessageReader>::CustomMessage>>, MessageHandlingError> {
-		log_trace!(self.logger, "Received message {:?} from {}", message, log_pubkey!(peer.their_node_id.unwrap()));
+		log_gossip!(self.logger, "Received message {:?} from {}", message, log_pubkey!(peer.their_node_id.unwrap()));
```
This should also be addressed by https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/fbbaeea3afd1f02da29c4b5e7981807b806973ff.

tnull commented:
Likewise very sorry for the delay from my side! I now addressed (most of) the comments and rebased.

TheBlueMatt commented:
Looking good, I think the only remaining thing is to convert some of the `ErrorAction::IgnoreAndLog(Level::Trace)` error references in `network_graph.rs` to `IgnoreAndLog(Level::Gossip)`.

tnull commented:
> Looking good, I think the only remaining thing is to convert some of the `ErrorAction::IgnoreAndLog(Level::Trace)` error references in `network_graph.rs` to `IgnoreAndLog(Level::Gossip)`.

Addressed in d5715200956cd346470db97328ddf99ac6fb04ed and rebased.

TheBlueMatt commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -757,7 +758,19 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
-		log_trace!(self.logger, "Enqueueing message {:?} to {}", message, log_pubkey!(peer.their_node_id.unwrap()));
+
+		match message.type_id() {
+			msgs::Ping::TYPE |
```
I don't think PING and PONG are gossip :)

tnull replied:
Yep, removed `Ping`/`Pong` and added `ChannelUpdate` with https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/753889b036d52eba9a95a57a610771d602cf8f74.

TheBlueMatt commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -757,7 +758,19 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
+		match message.type_id() {
+			msgs::Ping::TYPE |
+			msgs::Pong::TYPE |
+			msgs::ChannelAnnouncement::TYPE |
```
This appears to be missing ChannelUpdate.

jkczyz replied:
Give BOLT 1 defines "routing" messages are in a specific message type range, maybe we should add a function/method to `wire` to prevent missing any messages and including any new ones.

https://github.com/lightning/bolts/blob/master/01-messaging.md

tnull replied:
You're right, a function along the lines of `fn is_gossip_msg(type_id: u16)` would probably simplify this. However, this assumes that we strictly want to log all gossip messages via the `Gossip` level, and no other messages, even if they could spam the log output. Is this the case, then I'll gladly do it this way?

jkczyz replied:
Yeah, I think so. Maybe a function named `is_gossip` on `wire::Type` with a default implementation would be sufficient.

TheBlueMatt replied:
I dunno if we want to be super strict about it, but we could still unify the code - we probably actually want to log some of the filter/query stuff as TRACE since its not usually very noisy on "normal" connections. That said, we've never cared about trying to match LDK's interfaces up with the spec exactly, and dunno why we should start here. Maybe `is_gossip_msg` isn't the right name, maybe something like `is_gossip_update_msg` or something else to communicate its one of channel/node announcement/update messages and not something asking for such messages.

tnull replied:
As suggested, https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/e3b4d90ad981331270b3fb9dd1c7743e8f60efc2 implements `is_gossip_update_msg` on `wire::Type`, which simplifies and future-proves the code in `PeerHandler`.

jkczyz replied:
BOLT 1 calls the entire range as "routing" messages. Then in BOLT 7 the announcement and update messages are the ones considered "gossip" messages while the remaining are consider gossip queries as defined by the `gossip_queries` feature bit. @TheBlueMatt Are you saying we should log the former as level `Gossip` and the latter as `Trace` like all the other messages?

TheBlueMatt replied:
Yes, I think so. I don't really think there's any reason for this API to be tied to spec details. What we really want to do is make full trace logs not be GiB in size without removing any information that could potentially be useful. Queries could be important for debugging and also aren't very voluminous, so I suggested making the distinction "graph sync" messages, not just gossip ones.

jkczyz commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -757,7 +758,19 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
-		log_trace!(self.logger, "Enqueueing message {:?} to {}", message, log_pubkey!(peer.their_node_id.unwrap()));
+
+		match message.type_id() {
+			msgs::Ping::TYPE |
+			msgs::Pong::TYPE |
```
Should these be included given they are not gossip?

tnull replied:
Yep, removed `Ping`/`Pong` and added `ChannelUpdate` with https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/753889b036d52eba9a95a57a610771d602cf8f74.

jkczyz commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -953,7 +966,19 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 	) -> Result<Option<wire::Message<<<CMH as core::ops::Deref>::Target as wire::CustomMessageReader>::CustomMessage>>, MessageHandlingError> {
-		log_trace!(self.logger, "Received message {:?} from {}", message, log_pubkey!(peer.their_node_id.unwrap()));
+		match message {
+			wire::Message::Ping(_) |
+			wire::Message::Pong(_) |
```
Likewise?

tnull replied:
Yep, removed `Ping`/`Pong` and added `ChannelUpdate` with https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/753889b036d52eba9a95a57a610771d602cf8f74.

TheBlueMatt reviewed (commented):
LGTM, mod two comments.

jkczyz commented on lightning/src/util/logger.rs:
```diff
@@ -17,11 +17,13 @@
 /// An enum representing the available verbosity levels of the logger.
 #[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
 pub enum Level {
+	/// Designates extremely verbose information, including gossip-induced messages
+	Gossip,
 	/// Designates very low priority, often extremely verbose, information
 	Trace,
 	/// Designates lower priority information
```
We may want to consider category-based logging instead of level-based logging in the future. Or maybe both? No need to change here.

tnull replied:
Hum, but if I'm not mistaken a `Gossip` log level seems to be required for now to log gossip events separately?

jkczyz replied:
Yup, no need to change anything in this PR. Just pointing out that we may want to rethink this in the future.

tnull replied:
Great, thanks!

jkczyz reviewed (commented):
Largely looks good aside from the question on ping/pong logging. Could probably squash into one commit before merging.

tnull commented:
Addresses comments further and squashed commits in https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/753889b036d52eba9a95a57a610771d602cf8f74.

TheBlueMatt reviewed (commented):
Heh, I don't think you meant to check in `*/rusty-tags.vi`. Feel free to edit `.gitignore`.

tnull commented:
> Heh, I don't think you meant to check in `*/rusty-tags.vi`. Feel free to edit `.gitignore`.

Definitely not, very sorry about that. 😬😳

jkczyz commented on lightning/src/ln/wire.rs:
```diff
@@ -248,12 +283,27 @@ pub(crate) use self::encode::Encode;
 pub trait Type: core::fmt::Debug + Writeable {
 	/// Returns the type identifying the message payload.
 	fn type_id(&self) -> u16;
+	fn is_gossip_update_msg(&self) -> bool;
```
I was thinking you could just implement it directly here as a default implementation. BOLT 1 defines it over a message type range (256-511). That way we wouldn't accidentally forget to update the code below when a new gossip message is added.

TheBlueMatt replied:
I think I'd kinda prefer it default to not-gossip - even if its a gossip message, if its not something we see on the wire at high volume, we may want to log messages of it at TRACE and not GOSSIP.

TheBlueMatt replied:
In fact, I wouldn't object, though I don't feel strongly or care much, to changing the GossipTimestampFilter message to "not-gossip-update-message"

jkczyz replied:
In that case -- and given `wire` is a public module -- I think I'd prefer just adding a helper to `peer_handler` for determining what we'd like to consider gossip logging. Seems like an internal distinction that does not need to be exposed more broadly.

TheBlueMatt replied:
Fair enough, that feels like it may be more natural.

tnull replied:
Alright, I reverted the implementation on `wire::Type` and introduce a simple helper function with https://github.com/rust-bitcoin/rust-lightning/pull/1145/commits/625cf8308b55fc164e8aafb55a3fc22d20f79fe7

jkczyz commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -1611,6 +1599,20 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
+			msgs::ChannelAnnouncement::TYPE |
+			msgs::ChannelUpdate::TYPE |
+			msgs::NodeAnnouncement::TYPE |
+			msgs::QueryShortChannelIds::TYPE |
+			msgs::ReplyShortChannelIdsEnd::TYPE |
+			msgs::QueryChannelRange::TYPE |
+			msgs::ReplyChannelRange::TYPE |
+			msgs::GossipTimestampFilter::TYPE => true,
```
IIUC, we don't want these to be considered "gossip" for logging purposes. @TheBlueMatt Could you confirm?

TheBlueMatt replied:
Yea, I think we don't, probably better to leave them as TRACE for now and we can reduce them to GOSSIP later if its still too verbose. Probably only `Channel/NodeUpdate/Announcement`

tnull replied:
Okay, fixed with https://github.com/rust-bitcoin/rust-lightning/commit/7a2b716d26ae71b0020ab95b2ea3604f7f26c824.

tnull commented:
Further addressed comments, rebased, and squashed again.

jkczyz commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -1588,6 +1599,15 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 
+fn is_gossip_msg(type_id: u16) -> bool {
+		match type_id {
+			msgs::ChannelAnnouncement::TYPE |
+			msgs::ChannelUpdate::TYPE |
+			msgs::NodeAnnouncement::TYPE |
```
I think you want true for these three 😛

tnull replied:
True!

jkczyz commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -1588,6 +1599,15 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 }
 
+fn is_gossip_msg(type_id: u16) -> bool {
+		match type_id {
+			msgs::ChannelAnnouncement::TYPE |
+			msgs::ChannelUpdate::TYPE |
+			msgs::NodeAnnouncement::TYPE => true,
+			_ => false
+		}
```
Reduce indent by one.

tnull replied:
Done.

tnull commented:
> Will land after the trivial indent fix Jeff mentioned.

Fixed in https://github.com/rust-bitcoin/rust-lightning/commit/8913a5af4fec4865f7f2f111a0079eb89f05b96c. Thank you very much for the reviews!

TheBlueMatt reviewed (approved):
Will land after CI:

```diff
$ git diff-tree -U1 8913a5af4 3b4b74bc6
diff --git a/lightning/src/ln/peer_handler.rs b/lightning/src/ln/peer_handler.rs
index 5bb96f9cf..89f3e9ff5 100644
--- a/lightning/src/ln/peer_handler.rs
+++ b/lightning/src/ln/peer_handler.rs
@@ -1602,8 +1602,8 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 fn is_gossip_msg(type_id: u16) -> bool {
-		match type_id {
-			msgs::ChannelAnnouncement::TYPE |
-			msgs::ChannelUpdate::TYPE |
-			msgs::NodeAnnouncement::TYPE => true,
-			_ => false
-		}
+	match type_id {
+		msgs::ChannelAnnouncement::TYPE |
+		msgs::ChannelUpdate::TYPE |
+		msgs::NodeAnnouncement::TYPE => true,
+		_ => false
+	}
 }
$
```
